terraform roster

dmacvicar opened the pull request:
### What does this PR do?

Implements a dynamic roster from terraform current state file.

Currently supports libvirt and AWS resources.
(For libvirt, see https://github.com/dmacvicar/terraform-provider-libvirt)

For libvirt resources, currently the first address of the first interface is chosen.
For AWS resources, the public_dns or the public_ip attributes are used.

Example:

Given a simple salt-ssh tree with a Saltfile

```yaml
salt-ssh:
  config_dir: etc/salt
  max_procs: 30
  wipe_ssh: True
```

and etc/salt/master

```yaml
root_dir: .
  file_roots:
    base:
      - srv/salt
  pillar_roots:
    base:
      - srv/pillar
  roster: terraform
```

We setup a terraform file that creates N virtual machines (libvirt_domain resources) setup
with cloud-init and the salt-ssh ssh key. In this case we set N to 2.

```hcl
provider "libvirt" {
  uri = "qemu:///system"
}

resource "libvirt_cloudinit" "init" {
  name = "test-init.iso"
  ssh_authorized_key = "${file("etc/salt/pki/master/ssh/salt-ssh.rsa.pub")}"
}

resource "libvirt_volume" "opensuse_leap" {
  name = "opensuse_leap"
  source = "http://r.opensu.se/Cloud:Images:Leap_42.3/images/openSUSE-Leap-42.3-OpenStack.x86_64.qcow2"
}

resource "libvirt_volume" "volume" {
  name = "volume-${count.index}"
  base_volume_id = "${libvirt_volume.opensuse_leap.id}"
  count = 2
}

resource "libvirt_domain" "domain" {
  name = "domain-${count.index}"
  memory = 1024
  disk {
       volume_id = "${element(libvirt_volume.volume.*.id, count.index)}"
  }

  network_interface {
    network_name = "default"
    hostname = "minion${count.index}"
    wait_for_lease = 1
  }
  cloudinit = "${libvirt_cloudinit.init.id}"
  count = 2
}
```

```console
$ terraform apply
libvirt_cloudinit.init: Creating...
...
libvirt_volume.opensuse_leap: Creating...
...
libvirt_domain.domain.1: Creating...
libvirt_domain.domain.0: Creating...
...
libvirt_domain.domain.0: Creation complete
...
libvirt_domain.domain.1: Creation complete

Apply complete! Resources: 5 added, 0 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
 use the `terraform show` command.

State path: terraform.tfstate
```

You can start now using salt-ssh without any configuration.

```console
$ salt-ssh '*' test.ping
  domain-0:
    True
  domain-1:
    True
```

### What issues does this PR fix or reference?

None. I got inspired by an old post in the forum:
https://groups.google.com/forum/#!topic/salt-users/HMav5sk7zQ0

But that roster was never shared or upstreamed.

### Tests written?

Yes. One unit test (first roster with a test).

## Feedback/Review needed

I am looking for feedback on the following areas.

* I am not sure if the ssh key handling is right the best way to achieve a zero-configuration solution.
  (I kind of rely on setting the same key with cloud-init on the image)
* Right now for libvirt I just use the first ip. May be a ranking like other rosters could be a good idea, but not sure how to reuse code there.
* I had to refactor a bit flat roster in order to reuse it's matcher.

//cc @pkazmierczak @meaksh @dincamihai @isbm

dmacvicar committed:
initial terraform salt-ssh roster

isbm commented on salt/roster/__init__.py:
```diff
@@ -51,7 +51,7 @@ def get_roster_file(options):
 
     if not os.path.isfile(template):
-        raise IOError('No roster file found')
+        raise IOError('Roster file \'{0}\' not found'.format(template))
```
Maybe double-quotes instead of back-slashes?
python
```
"Roster file '{0}' not found". format(...
```

isbm commented on salt/roster/__init__.py:
```diff
@@ -104,6 +104,7 @@ def targets(self, tgt, tgt_type):
             except salt.exceptions.SaltRenderError as exc:
                 log.error('Unable to render roster file: {0}'.format(exc))
             except IOError as exc:
+                log.error('Can\'t access roster for backend {0}: {1}'.format(back, exc))
```
Single quotes aren't mandatory. And if you have them inside, please use double-quotes instead.

dmacvicar replied:
Salt conventions do indicate use of single quotes:

> In Salt, all strings use single quotes unless there is a good reason not to. This means that docstrings use single quotes, standard strings use single quotes etc.:

So, this boils down on whether to use double quotes as an exception here and why that is better than a backslash. Can you point me to any resource indicating why double quotes is preferred to a backslash in this case, or are you just pushing your personal preference?

isbm replied:
It is @thatch45 that hates double-quotes. :-) But I think he will agree that `"Fred's bar"` is better than `'Fred\'s bar`. And we still have plenty of [similar cases](https://github.com/saltstack/salt/blob/develop/salt/transport/raet.py#L96).

dmacvicar replied:
changed

isbm commented on salt/roster/flat.py:
```diff
@@ -44,19 +44,20 @@ def targets(tgt, tgt_type='glob', **kwargs):
     Matcher for the roster data structure
     '''
-    def __init__(self, raw, tgt, tgt_type, ipv='ipv4'):
+    def __init__(self, raw, tgt, tgt_type, ipv='ipv4', opts={}):
```
This is not pythonic way. This is:
```python
def __init__(self, raw, tgt, tgt_type, ipv='ipv4', **opts):
```

dmacvicar replied:
Ok.

Can you please you are sure about this. The code above has also kwargs, but here we are talking about passing global `__opts___` to avoid passing it as a global. If I use `kwargs` where you indicate, I would end merging `__opts__` into the function `kwargs`, which I am not sure it is better.

isbm replied:
Ah, right, I noticed now what you want to do. I think it depends where you reuse the class or not in the future. If you plan to call it:
```python
rmatcher = RosterMatcher(raw, tgt, tgt_type, 'ipv6', __opts__)
```
Then your way is would work. However personally I would go the kwargs way as it accepts _only_ named arguments and thus do this:
```python

def __init__(self, raw, tgt, tgt_type, ipv='ipv4', **kwargs):
    self.opts = kwargs.get('opts', {}).copy()
```
In either way you still need to fix a bug, as in your case you create only a reference to an `__opts__`, and so if you change anything in `self.opts`, you also change inside `__opts__`. IIRC, this is not what you want (otherwise you'd better_just_ access `__opts__` directly to avoid confusion).

dmacvicar replied:
I think you are mixing two different problems 1) re-usability and local variables 2) immutability.

* Right now the component _is_ accessing `__opts__` directly, so the component does not assume it is immutable. I don't see why I should change that.
* However, the component depends on a global, which means I can't use it in different contexts, which is what I intend to do.

So I still think

```python
class RosterMatcher(object):
    def __init__(self, raw, tgt, tgt_type, ipv='ipv4', opts={}):
    ...
```
and
```python
matcher = RosterMatcher(raw, tgt, tgt_type, 'ipv6', __opts__)
```
is the best option that solves that problem. If that is bad style for _some_ reason, then

```python
class RosterMatcher(object):
    def __init__(self, raw, tgt, tgt_type, ipv='ipv4', **kwargs):
    ...
```
and
```python
matcher = RosterMatcher(raw, tgt, tgt_type, 'ipv6', **__opts__)
```

Which I don't like because it does an implicit copy, which is unrelated to my problem.

I can just pass __opts__ as a kwargs itself like you describe above, without the copy.

```python
class RosterMatcher(object):
    def __init__(self, raw, tgt, tgt_type, ipv='ipv4', **kwargs):
        self.opts = kwargs.get('opts', {})
    ...
```
and
```python
matcher = RosterMatcher(raw, tgt, tgt_type, 'ipv6', opts=__opts__)
```
Which is the one I like more.

dmacvicar replied:
changed

isbm commented on salt/roster/flat.py:
```diff
@@ -129,7 +130,7 @@ def ret_range_minions(self):
 
         minions = {}
-        range_hosts = _convert_range_to_list(self.tgt, __opts__['range_server'])
+        range_hosts = _convert_range_to_list(self.tgt, self.opts['range_server'])
```
Just a note: then make sure that `range_server` key is always in the `opts`. Otherwise this will crash.

dmacvicar replied:
This is an unrelated fix to the PR.

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+    name = attrs['name']
+    num_ifaces = int(attrs['network_interface.#'])
+    for if_k in range(num_ifaces):
+        num_addrs = int(attrs['network_interface.{}.addresses.#'.format(if_k)])
```
A nitpick: in Salt we don't do empty braces for formatting, but at least one index (or name). I remember Pylint was also crying about this at least times ago.

dmacvicar replied:
fixed

cachedout replied:
One note: we are finally ready to drop 2.6 support and we have started accepting non-2.6 code into develop. :]

isbm replied:
@cachedout great news, BTW!

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+                continue
+            ret[name] = {'host': addr}
+            _add_ssh_key(ret[name])
+            return
```
Not needed statement.

dmacvicar replied:
Not correct. If you remove `return` you will continue looping.

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+        for addr_k in range(num_addrs):
+            addr = attrs['network_interface.{}.addresses.{}'.format(if_k, addr_k)]
+            if not addr:
+                continue
```
I would `if addr:` then (things below) instead.

dmacvicar replied:
Thanks, but I don't agree with that style. I prefer to follow the pattern of _exiting or bailing out early_, get the bad cases out of the way, and _avoiding_nesting_.

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+        return
+
+    ret[name] = {'host': addr}
+    _add_ssh_key(ret[name])
```
Two returns in the middle is a bad style in Python. I would replace the whole thing like this:
```python
def _handle_aws_resource(resource, ret):
    attrs = resource.get('primary', {}).get('attributes', {})
    addr = attrs.get('public_dns', attrs.get('public_ip'))
    name = attrs.get('tags.Name', attrs.get('id'))
    if name and addr:
        ret[name] = {'host': addr}
        _add_ssh_key(ret[name])
```

dmacvicar replied:
Ok, I disagree with the return vs nesting, but I like the idea of using defaults in `.get()`

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+    '''
+    priv = None
+    if __opts__.get('ssh_use_home_key') and \
+       os.path.isfile(os.path.expanduser('~/.ssh/id_rsa')):
```
Note: It is OK to go 120 chars for a line here.

dmacvicar replied:
changed

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+    if not priv:
+        return
+    if os.path.isfile(priv):
+        ret['priv'] = priv
```
Maybe drop that return in the middle? It is better:
```python
if priv and os.path.isfile(priv):
    ret['priv'] = priv
```

dmacvicar replied:
yes, here it makes sense, the early bailout does not simplify anything.

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+    with salt.utils.fopen(state_file_path, 'r') as fh_:
+        tfstate = json.load(fh_)
+
+    modules = tfstate.get('modules', None)
```
The `.get(..)` will give you `None` by default anyway, so you don't need to be redundant.

dmacvicar replied:
fixed

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+
+    if not modules:
+        log.error('Malformed tfstate file. No modules found')
+        return ret
```
I would not make `modules` as `None` for an extra `return` in the middle, but make it iterable (`dict` in your case). So the loop below will skip it.

dmacvicar replied:
again, I find indenting the default case a bad style. _bail out early_.

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+
+    for module in modules:
+        resources = module['resources']
+        for _, resource in six.iteritems(resources):
```
Throwaway variables are better avoided. How about:
```python
for resource in resources.values():
...
```

...instead?

dmacvicar replied:
fixed

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+    Returns the roster from the terraform state file, checks opts for location, but defaults to terraform.tfstate
+    '''
+    roster_file = os.path.abspath('terraform.tfstate')
+    if __opts__.get('roster_file', None):
```
No need to extra ask for 'None'. Also it is correct this way:
```python
if __opts__.get('roster_file') is None:
    ...
```

dmacvicar replied:
fixed

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+        roster_file = os.path.abspath(__opts__['roster_file'])
+
+    if not os.path.isfile(roster_file):
+        log.error('Can\'t find terraform state file {}'.format(roster_file))
```
Please use double-quotes and format with an index (no empty `{}`, since they won't work on Python 2.6).

dmacvicar replied:
fixed

isbm commented on salt/roster/terraform.py:
```diff
@@ -0,0 +1,237 @@
+        log.error('Can\'t find terraform state file {}'.format(roster_file))
+        return {}
+
+    log.debug('terraform roster: using {} state file'.format(roster_file))
```
Empty formatting.

dmacvicar replied:
fixed

isbm commented on tests/unit/roster/terraform.py:
```diff
@@ -0,0 +1,51 @@
+
+if __name__ == '__main__':
+    import unittest
+    unittest.main()
```
This whole block is not really needed.

dmacvicar replied:
It is if you want to run that isolated test. At least the [documented way]((https://docs.saltstack.com/en/latest/topics/development/tests/index.html#running-test-subsections) is not working for me.

```console
$ python tests/runtests.py -n unit.roster.terraform
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Python Version: 2.7.13 (default, Jan 03 2017, 17:41:54) [GCC]
 * Transplanting configuration files to '/tmp/salt-tests-tmpdir/config'
 * Current Directory: /space/git/salt/salt
 * Test suite is running under PID 31215
 * Logging tests on /tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Could not locate test 'unit.roster.terraform'. Exiting.
```

s0undt3ch replied:
Even with the latest changes to develop it should work, will try to see why it doesn't.
Meanwhile, does the following work?

```
python tests/runtests.py -n tests.unit.roster.terraform
```

Ah! It all makes sense now, please rename your test file to `test_terraform.py`.
All test modules must now be prefixed with `test_`, but the test suit is probably warning you about it already.

isbm replied:
OK, so _not_ needed.

dmacvicar replied:
No.

I have tried all combinations of being in `salt/` or `salt/tests` and doing:
(I also tried renaming terraform.py to test_terraform.py)

```console
python tests/runtests.py -n tests.unit.roster.test_terraform
python tests/runtests.py -n unit.roster.test_terraform
python tests/runtests.py -n tests.unit.roster.terraform
python tests/runtests.py -n unit.roster.terraform

python ./runtests.py -n unit.roster.test_terraform
python ./runtests.py -n unit.roster.test_terraform
python ./runtests.py -n unit.roster.terraform
python ./runtests.py -n unit.roster.terraform
```

Interestingly, 

```console
python tests/runtests.py -n unit.beacons.test_adb_beacon
```
works...

You can also see that running with `unit...` vs `tests.unit...` has different effects, and the daemons get started in the second case. So it is not just about resolving the path.

dmacvicar replied:
Solved. Missing empty `__init__.py` in `unit/roster`.

s0undt3ch replied:
Dam, missed the missing `__init__.py`.

As for the test daemons starting when using `-n tests.unit.X`, that's missed logic in the latest changes that I need to fix.

s0undt3ch replied:
You can also do, with the latest changes in the develop branch:

```
python tests/runtests.py tests/unit/roster/test_terraform.py
```

isbm replied:
@s0undt3ch why not PyTest, BTW? I just tried @dmacvicar's PR:
```
$ pytest unit/roster/test_terraform.py 
========================  test_terraform.py  =========================
..
----------------------------------------------------------------------
Ran 2 tests in 0.005s

OK
*******************************************************************************
```

s0undt3ch replied:
Pytest works if you run a single test module or only the integration tests or probably only the unittests(but these will fail for sure). Once you mix unittest with integration tests it will fail badly. The reason is that salt's dunders are being populated at the module level instead of being patched by mock on `setUp` which has all kinds of secondary effects.

However, pytest is the goal, but a lot of cleaning up must happen before we completely move to pytest, unfortunately.

isbm replied:
@s0undt3ch Does it [really](github.com/openSUSE/salt-toaster) fail?

s0undt3ch replied:
Still haven't tried your toaster work, but just test this:
```
pytest -ra -v tests/
```

isbm replied:
@s0undt3ch naw, not _like that_ of course. There are few little things needed to be added. But overall, it works like a charm.

isbm commented on tests/unit/roster/terraform.py:
```diff
@@ -0,0 +1,51 @@
+        tfstate = os.path.join(os.path.dirname(__file__), 'terraform.data', 'terraform.tfstate')
+        terraform.__opts__['roster_file'] = tfstate
+        pki_dir = os.path.abspath(
+            os.path.join(os.path.dirname(__file__), 'terraform.data'))
```
Don't have to wrap here and better don't.

dmacvicar replied:
fixed

isbm reviewed (changes requested):
@dmacvicar please take a look at these. Thanks!

dmacvicar commented:
@isbm Everything mentioned should be addressed by now.

I am looking forward for feedback from others about the points mentioned in the summary.

dubb-b commented:
@dmacvicar Could you dynamically build a variables.tf file that has the ssh key in there and then use the Terraform Variables to set the "ssh_authorized_key" value? Then when you run terraform plan/apply it will grab it from there.

dmacvicar commented:
@dubb-b can you describe the workflow in detail? How will be the terraform variables files be generated (the machines are not up at that point)?

s0undt3ch commented:
Yep, the test suite is warning you about the rename required.

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/9101/testReport/junit/unit.test_test_module_names/BadTestModuleNamesTestCase/test_module_name/

isbm commented:
@dmacvicar [some little fixes](https://github.com/dmacvicar/salt/pull/1) while @dubb-b describes the workflow.

gtmanfred commented on salt/roster/flat.py:
```diff
@@ -44,19 +44,20 @@ def targets(tgt, tgt_type='glob', **kwargs):
         self.tgt_type = tgt_type
         self.raw = raw
         self.ipv = ipv
+        self.opts = kwargs.get('opts', {})
```
shouldn't this set `self.opts` to `__opts__` if `kwargs['opts']` doesn't exist?

dmacvicar replied:
It could. Depends if we want to make it crash if the global is not there.

gtmanfred replied:
IMO, like with pretty much everything else, the __opts__ should always be there, because it should only be loaded from the loader.

dubb-b commented:
@dmacvicar 

You are right. I was thinking about the intent differently. My suggestion would not help now that I understand what you are doing more clearly.

dmacvicar commented:
@dubb-b I think generating a variables.tf from Salt would result in a chicken-egg problem.

What about making the roster module read a certain output variable, eg. `salt-ssh.key`, and then you could in your terraform file do something like:

```hcl
resource "libvirt_cloudinit" "init" {
  name = "test-init.iso"
  ssh_authorized_key = "${file("/home/duncan/mykey.pub")}"
}

// ...

resource "libvirt_domain" "domain" {
  name = "domain-${count.index}"
  memory = 1024
  disk {
       volume_id = "${libvirt_volume.volume.1.id}"
  }

  network_interface {
    network_name = "default"
    hostname = "minion"
    wait_for_lease = 1
  }
  cloudinit = "${libvirt_cloudinit.init.id}"
}

output "salt-ssh.key" {
  value = "${init.ssh_authorized_key}"

}
```

The roster will use that output variable if present, or fallback to a default (eg. `etc/salt/pki/master/ssh/salt-ssh.rsa.pub`) if not there.

What do you think=

dubb-b commented:
Thats perfect, I was thinking of this issue in a much different way. After I looked more at the code and what you were doing I understood why I was way out in left field. 

The outputs scenario would definitely work. 

What if you use different keys for different parts of the infrastructure(e.g dmz.key vs trustzone.key)? 

May have to figure out a way to determine which key to use for which resource. Although some people don't use more than one key.

dmacvicar commented:
@dubb-b I was thinking that this could be extended not only to the ssh-key but also every roster attribute.

It can be hierarchical:

* Without any override, defaults will be used. That is, very provider dependent: AWS instances use `public_ip`. For ssh keys we use the salt default (`pki_dir + salt-ssh + master + salt-ssh.pub`).

* If an output variable of the form `salt-ssh.rostervar` is found. It is used for all hosts. Example:

```hcl
output "salt-ssh.priv" {
  value = "/home/me/.ssh/id_rsa"
}
```

* unless an output variable of the form `salt-ssh.resourcename.rostervar` or `resourcename.salt-ssh.rostervar` (????) is used, then the value is overridden only for that resource.

```hcl
output "libvirt_domain.domain.0.salt-ssh.priv" {
  value = "/home/me/.ssh/id_rsa"
}
```

What do you think?

dubb-b commented:
@dmacvicar 

That should work out, then you can use that same hierarchical model for all the rest of the roster attributes as well.  I have thinking about any caveats and if I come up with any Ill post back. Thanks for taking this on. I want to give it a try soon. I'll post back with my testing.

cachedout commented:
@dubb-b Did you get a chance to test this?

dmacvicar commented:
@cachedout I will prefer to keep this [WIP] until I implement the configuration overrides discussed [here](https://github.com/saltstack/salt/pull/39754#issuecomment-283991687). Just need to find some time to do it.

dubb-b commented:
@dmacvicar Just let me know when you are ready to test. I will give it a shot.

cachedout commented:
@dmacvicar Sounds good. Thanks for the update.

dmacvicar commented:
@cachedout Can I use RestructuredText in the docstring of the module or should I then put all the "long" doc in `doc/ref/roster/all/salt.roster.terraform.rst` ?

cachedout commented:
@dmacvicar You can use it in the docstring and Sphinx should pick it up when the docs are built.

dmacvicar commented:
I have implemented overriding values. As documented in the module:

# Overriding values

An output variable named `salt.roster.VARNAME` in a module would override the `VARNAME` roster
attribute for all resources of that module.

.. code-block:: hcl

    output "salt.roster.priv" {
      value = "/path/to/id_rsa"
    }

An output variable named `salt.roster.VARNAME.RESOURCENAME` in a module would override the `VARNAME` roster
attribute for resource `RESOURCENAME` of that module.

.. code-block:: hcl

    output "salt.roster.priv.libvirt_domain.domain.1" {
      value = "/path/to/id_rsa"
    }

I also squashed all the commits, so I may need to check lint again. I can't run lint locally again. @dubb-b

dubb-b commented:
@dmacvicar Let me get a terraform multi instance setup going. Are you at the point you want me to go ahead and test it out?  I can get a Multi Instance setup in AWS with Terraform and give it a shot.

holmboe commented:
I am eagerly awaiting this merge so that I can add Digital Ocean support for it. Cheering on the sidelines here.

dmacvicar commented:
@dubb-b I am relying a lot on the unit test I wrote for the new overriding functionality, I did not test extensively outside of them because I do not have a personal usecase for it.

@holmboe that would be awesome! If you need any help setting up a dev environment or do not understand my code, just let me know!

cachedout commented:
I would really like to get this in. @dmacvicar it sounds like we weren't able to get the requested testing done but are you OK with getting this merged anyhow?

dubb-b commented:
@dmacvicar I am getting this error: [ERROR   ] No matching targets found in roster.

It may be my setup, can you update the instructions to add more details about where to place the tfstate file and your Saltfile?  Just in case i'm doing this wrong.

cachedout commented:
@dmacvicar Did you see the above from @dubb-b ?

dmacvicar commented:
@dubb-b I will try to push a working environment to my github with the libvirt configuration I am using.

Could you please, show/tell me

* how are you running `salt-ssh`? Is in a single contained directory or a system-wide installation?
* Are you running salt-ssh in the same directory where terraform leaves its `terraform.tfstate` file?
* If you add `-l debug` to `salt-ssh` do you see any other warnings or messages?
* Are you using it with libvirt or AWS? can you include or send me your `terraform.tfstate` file? (you can change names and ips if you want).

Thanks!

dubb-b commented:
@dmacvicar I used AWS, I believe that I still have the node and its shutdown. Ill fire/terraform apply it and give you the information that you requested.

Today I am tied up, I will update you tomorrow AM MT.

Thanks!

dubb-b commented:
salt-ssh is being ran in the same directory.

```salt-ssh '*' test.ping -l debug
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: salt-master01-dev.c7.saltstack.net
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] LazyLoaded terraform.targets
[DEBUG   ] terraform roster: using /tmp/roster/salt/terraform.tfstate state file
[WARNING ] terraform: No adress for resource 'web01' in state
[WARNING ] terraform: No adress for resource 'web02' in state
[DEBUG   ] 0 hosts in terraform state file
[DEBUG   ] Matched minions: {}
[DEBUG   ] LazyLoaded roots.envs
[DEBUG   ] Could not LazyLoad roots.init: 'roots.init' is not available.
[DEBUG   ] Updating roots fileserver cache
[DEBUG   ] LazyLoaded local_cache.prep_jid
[DEBUG   ] Adding minions for job 20170512163952732998: []
[ERROR   ] No matching targets found in roster.
```
Dir Structure: the tfstate file is in the salt dir...

```
-rw-r--r--  1 root root       828 May 12 15:10 main.tf
drwxr-xr-x 13 root root      4096 May 12 16:39 salt
-rw-r--r--  1 root root        69 May 12 15:33 SaltFile
-rwxr-xr-x  1 root root 125085344 May 12 15:30 terraform
-rw-r--r--  1 root root       944 May 12 15:12 variables.tf
```
Here is one of the nodes of the tfstate file:

``` "primary": {
                        "id": "i-058589b881d804b64",
                        "attributes": {
                            "ami": "ami-8ca83fec",
                            "associate_public_ip_address": "true",
                            "availability_zone": "us-west-2a",
                            "disable_api_termination": "false",
                            "ebs_block_device.#": "0",
                            "ebs_optimized": "false",
                            "ephemeral_block_device.#": "0",
                            "iam_instance_profile": "",
                            "id": "i-058589b881d804b64",
                            "instance_state": "running",
                            "instance_type": "t2.micro",
                            "ipv6_address_count": "0",
                            "ipv6_addresses.#": "0",
                            "key_name": "brett",
                            "monitoring": "false",
                            "network_interface.#": "0",
                            "network_interface_id": "eni-906614bb",
                            "primary_network_interface_id": "eni-906614bb",
                            "private_dns": "ip-172-30-0-82.us-west-2.compute.internal",
                            "private_ip": "172.30.0.82",
                            "public_dns": "",
                            "public_ip": "x.x.x.x",
                            "root_block_device.#": "1",
                            "root_block_device.0.delete_on_termination": "true",
                            "root_block_device.0.iops": "100",
                            "root_block_device.0.volume_size": "8",
                            "root_block_device.0.volume_type": "gp2",
                            "security_groups.#": "0",
                            "source_dest_check": "true",
                            "subnet_id": "subnet-35f95f50",
                            "tags.%": "1",
                            "tags.Name": "web02",
                            "tenancy": "default",
                            "volume_tags.%": "0",
                            "vpc_security_group_ids.#": "1",
                            "vpc_security_group_ids.2274062738": "sg-88a865fb"
                        },
                        "meta": {
                            "schema_version": "1"
                        },
                        "tainted": false
                    },
                    "deposed": [],
                    "provider": ""
                }
            },
            "depends_on": []
        }
    ]
}
```
terraform version : `Terraform v0.9.5`

Let me know if you need more info, Ill keep these servers around.

dubb-b commented:
@dmacvicar 
 `"public_dns": "",`
is null for me because I am in a private subnet with a NATGW and no access to the public internet.
Some people may have VPN's attached and use the private interface, was this intended for that?

I am going to change VPC's and subnets and retest.
Here is the line I found:
` addr = attrs.get('public_dns', attrs.get('public_ip'))`

Ill let you know what I find...

dubb-b commented:
@dmacvicar 

Works! This is great!

```[root@salt-master01-dev roster]# salt-ssh '*' test.ping --user=ec2-user
web02:
    True
web01:
    True
```
My issue is that I dont have a publicDNS so i just added the IP in the terrafrom.tfstate file and passed the right user. And it worked great!

cachedout commented:
Go Go Jenkins!

dmacvicar commented:
@dubb-b Do you see a way to support non-public ip-addresses? or may be we should just document it?

dubb-b commented:
@dmacvicar I would try adding a parameter for public vs private and it uses whichever it finds first. Default to public. It's going to have to try and get a DNS and IP. In my case I have a Public IP but no DNS, so you could just try both, if nothing is found throw an error. I think having the code try both options would be great!

cachedout commented:
@dubb-b Could you be more specific?

dubb-b commented:
@cachedout 
Can he add command line params to salt-ssh for the roster. Then with this you could pass in something like `salt-ssh '*' test.ping --user=ec2-user --address=private --useIP=true`

`--user=ec2-user --address=private --useIP=true`
or 
`--user=ec2-user --address=public --useIP=false`

Something to allow the user to specify his needs.

cachedout commented:
@dubb-b I would prefer that this be passed via something like pillar options.

cachedout commented:
@dmacvicar Any thoughts on the above?

dmacvicar commented:
@cachedout I don't have a good concept yet for configuring public/private_ips. Given that it is documented AWS takes the ip from `public_ip/dns` I don't see this as a problem, we can add the functionality later.

Also, given that you can override any roster value using terraform output variables, I believe you could do something like (overriding `host` for one AWS resource):

```hcl
    output "salt.roster.host.aws_instance.instance.1" {
      value = "${aws_instance.instance.1.private_ip}"
    }
```
(I haven't tested this).

Maybe when others add support for more resources (eg. Digital Ocean), we come up with a better way.

cachedout commented:
I left some minor comments in-inline. Other than that, though, I think we're ready to go with this.

Commit:
Merge branch 'develop' into dmacvicar_terraform_roster

dmacvicar commented:
I did not see any AI for me.... let me know if you need anything!

cachedout commented:
@dmacvicar Could you please fix the final lint errors on here? https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/12437/violations/file/salt/roster/terraform.py/

cachedout commented:
@dmacvicar We would still really like to get this in. Any chance you can resolve the merge conflict here?

Commit:
Merge branch 'develop' into dmacvicar_terraform_roster

Commit:
Merge branch 'develop' into dmacvicar_terraform_roster

rallytime commented:
There are some lint violations here: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/15099/violations/

@dmacvicar Can you swing by and fix those?

Commit:
Merge branch 'develop' into dmacvicar_terraform_roster

dmacvicar committed:
Merge branch 'develop' into dmacvicar_terraform_roster

dmacvicar committed:
Merge branch 'develop' into dmacvicar_terraform_roster

dmacvicar committed:
Use self.opts instead of __opts__ which does not exist

dmacvicar committed:
Replace deprecated 'salt.utils.fopen' with 'salt.utils.files.fopen'

dmacvicar committed:
Make pylint happy

dmacvicar committed:
Make pylint happy

dmacvicar committed:
Port testcase to LoaderModuleMockMixin

dmacvicar committed:
Merge branch 'develop' into dmacvicar_terraform_roster

rallytime commented:
@dmacvicar I see you've been pushing some commits up to this PR lately. Is this ready to be reviewed again?

rallytime commented:
re-run py3

rallytime commented:
@dmacvicar Ah, I am making my way through some older PRs today and it looks like this one has some merge conflicts now. Can you rebase and fix those when you get a moment?

rallytime commented:
@dmacvicar Is this something you're still working on? Or should we close this for now and you can open it up again later when you're ready to keep working on it?

rallytime commented:
Since we haven't heard back on this for a while, I am going to close it. When you're ready, let us know and we can open this again. Or you can open a new PR, whichever is easier for you. :)

Thanks!

epcim commented:
@rallytime, I think this PR is a missing gap. It's needed in the ecosystem unless some plans for salt.cloud to support terraform. Can you please rebase (if clean) and re-run tests + provide a link to all violations.

@dmacvicar what is your interest for this rooster today? Experience from your personal usage? Is it worth to continue?

dmacvicar commented:
@epcim 

* There was a good reason/limitation about using output. I can't recall it anymore.
* Output variables can be used to pass values to the roster. A concept called "overriding" that emerged during the discussions.

About merging it:

* Too much time has happened between this was opened for first time.
* I tried to be proactive to include support even for parts I don't use much, for example AWS instances.
My motivation came from libvirt to use it with my [provider](https://github.com/dmacvicar/terraform-provider-libvirt).
* I don't have a Salt development environment at hand, and the loop until now has not been very efficient, I kind of lost motivation of working on it for a bigger audience.
* Looking at the discussions, I realize I would not be able to maintain it properly and handle wishes from everyone, I would prefer to close the issue.

dmacvicar commented:
I will propose a different solution for this, but I will go through an RFC.
